# Daten einlesen

COVID-19 Data Repository by the Center for Systems Science and Engineering (CSSE) at Johns Hopkins University
https://github.com/CSSEGISandData/COVID-19

Link zu den Daten bekommen:
- Datei raussuchen: 
    - **COVID-19** / **csse_covid_19_data** / **csse_covid_19_time_series** / 
- Datei `time_series_covid19_confirmed_global.csv` anklicken und Link von `RAW` (rechts oben) kopieren und diesen verwenden


Einlesen analog Vorgehen hier: https://towardsdatascience.com/covid-19-data-processing-58aaa3663f6

## `confirmed` einlesen

In [25]:
import pandas as pd


data_confirmed="https://github.com/CSSEGISandData/COVID-19/raw/master/csse_covid_19_data/csse_covid_19_time_series/time_series_covid19_confirmed_global.csv"

df_confirmed_wide = pd.read_csv(data_confirmed)
df_confirmed_wide

Province/State      Country/Region        Lat        Long  1/22/20  \
0              NaN         Afghanistan  33.939110   67.709953        0   
1              NaN             Albania  41.153300   20.168300        0   
2              NaN             Algeria  28.033900    1.659600        0   
3              NaN             Andorra  42.506300    1.521800        0   
4              NaN              Angola -11.202700   17.873900        0   
..             ...                 ...        ...         ...      ...   
268            NaN             Vietnam  14.058324  108.277199        0   
269            NaN  West Bank and Gaza  31.952200   35.233200        0   
270            NaN               Yemen  15.552727   48.516388        0   
271            NaN              Zambia -13.133897   27.849332        0   
272            NaN            Zimbabwe -19.015438   29.154857        0   

     1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  ...  1/17/21  1/18/21  \
0          0        0        0        0        0  ...    53984    54062   
1          0        0        0        0        0  ...    67690    67982   
2          0        0        0        0        0  ...   103833   104092   
3          0        0        0        0        0  ...     9083     9083   
4          0        0        0        0        0  ...    18875    18926   
..       ...      ...      ...      ...      ...  ...      ...      ...   
268        2        2        2        2        2  ...     1537     1539   
269        0        0        0        0        0  ...   152031   152555   
270        0        0        0        0        0  ...     2112     2113   
271        0        0        0        0        0  ...    37605    38207   
272        0        0        0        0        0  ...    27203    27892   

     1/19/21  1/20/21  1/21/21  1/22/21  1/23/21  1/24/21  1/25/21  1/26/21  
0      54141    54278    54403    54483    54559    54595    54672    54750  
1      68568    69238    69916    70655    71441    72274    72812    73691  
2     104341   104606   104852   105124   105369   105596   105854   106097  
3       9194     9308     9379     9416     9499     9549     9596     9638  
4      19011    19093    19177    19269    19367    19399    19476    19553  
..       ...      ...      ...      ...      ...      ...      ...      ...  
268     1540     1544     1546     1548     1548     1548     1549     1551  
269   153093   153590   154063   154557   155006   155414   155884   156393  
270     2115     2115     2115     2118     2118     2118     2118     2119  
271    39515    40949    42213    43333    44592    45337    46146    47622  
272    28675    29408    30047    30523    31007    31320    31646    32004  

[273 rows x 375 columns]

Die Daten liegen mit jedem Datum pro Spalte vor. Diese Art ist für die Verarbeitung mit `Pandas` sehr ungeeignet, weshalb wir aus dem **breiten** (nach rechts) Dataframe jetzt einen **langen** (nach unten) erstellen und nutzen dafür den `.melt` Befehl

In [26]:
# Liste der Einträge aller Daten erstellen 
dates = df_confirmed_wide.columns[4:] # Nimm alle Spalten ab der fünften (0=erste Spalte)
dates

Index(['1/22/20', '1/23/20', '1/24/20', '1/25/20', '1/26/20', '1/27/20',
       '1/28/20', '1/29/20', '1/30/20', '1/31/20',
       ...
       '1/17/21', '1/18/21', '1/19/21', '1/20/21', '1/21/21', '1/22/21',
       '1/23/21', '1/24/21', '1/25/21', '1/26/21'],
      dtype='object', length=371)

In [27]:
df_confirmed_long = df_confirmed_wide.melt(
    id_vars=['Province/State', 'Country/Region', 'Lat', 'Long'], 
    value_vars=dates, 
    var_name='Date', 
    value_name='Confirmed'
)
df_confirmed_long

Province/State      Country/Region        Lat        Long     Date  \
0                 NaN         Afghanistan  33.939110   67.709953  1/22/20   
1                 NaN             Albania  41.153300   20.168300  1/22/20   
2                 NaN             Algeria  28.033900    1.659600  1/22/20   
3                 NaN             Andorra  42.506300    1.521800  1/22/20   
4                 NaN              Angola -11.202700   17.873900  1/22/20   
...               ...                 ...        ...         ...      ...   
101278            NaN             Vietnam  14.058324  108.277199  1/26/21   
101279            NaN  West Bank and Gaza  31.952200   35.233200  1/26/21   
101280            NaN               Yemen  15.552727   48.516388  1/26/21   
101281            NaN              Zambia -13.133897   27.849332  1/26/21   
101282            NaN            Zimbabwe -19.015438   29.154857  1/26/21   

        Confirmed  
0               0  
1               0  
2               0  
3               0  
4               0  
...           ...  
101278       1551  
101279     156393  
101280       2119  
101281      47622  
101282      32004  

[101283 rows x 6 columns]

- der neue lange Dataframe `df_long` hat **100191 rows × 6 columns** = 601.146 Einträge
- der alte breite Dataframe `df_wide` hat **273 rows × 371 columns** = 101.283 Einträge

Die Einträge haben sich also für unseren neuen Dataframe `df_long` vervielfacht, da jetzt das Datum immer dazu geschrieben werden muss und somit oft doppelt vorkommt

### `deaths` + `recovered` analog 

In [28]:
data_death="https://github.com/CSSEGISandData/COVID-19/raw/master/csse_covid_19_data/csse_covid_19_time_series/time_series_covid19_deaths_global.csv"
data_recovered="https://github.com/CSSEGISandData/COVID-19/raw/master/csse_covid_19_data/csse_covid_19_time_series/time_series_covid19_recovered_global.csv"


df_death_wide = pd.read_csv(data_death)
df_recovered_wide = pd.read_csv(data_recovered)

df_death_long = df_death_wide.melt(
    id_vars=['Province/State', 'Country/Region', 'Lat', 'Long'], 
    value_vars=dates, 
    var_name='Date', 
    value_name='Deaths'
)

recovered_df_long = df_recovered_wide.melt(
    id_vars=['Province/State', 'Country/Region', 'Lat', 'Long'], 
    value_vars=dates, 
    var_name='Date', 
    value_name='Recovered'
)


#### alle drei Dataframes zusammenführen

In [29]:
# Merging confirmed_df_long and deaths_df_long
full_table = df_confirmed_long.merge(
  right=df_death_long, 
  how='left',
  on=['Province/State', 'Country/Region', 'Date', 'Lat', 'Long']
)
# Merging full_table and recovered_df_long
full_table = full_table.merge(
  right=recovered_df_long, 
  how='left',
  on=['Province/State', 'Country/Region', 'Date', 'Lat', 'Long']
)

full_table

Province/State      Country/Region        Lat        Long     Date  \
0                 NaN         Afghanistan  33.939110   67.709953  1/22/20   
1                 NaN             Albania  41.153300   20.168300  1/22/20   
2                 NaN             Algeria  28.033900    1.659600  1/22/20   
3                 NaN             Andorra  42.506300    1.521800  1/22/20   
4                 NaN              Angola -11.202700   17.873900  1/22/20   
...               ...                 ...        ...         ...      ...   
101278            NaN             Vietnam  14.058324  108.277199  1/26/21   
101279            NaN  West Bank and Gaza  31.952200   35.233200  1/26/21   
101280            NaN               Yemen  15.552727   48.516388  1/26/21   
101281            NaN              Zambia -13.133897   27.849332  1/26/21   
101282            NaN            Zimbabwe -19.015438   29.154857  1/26/21   

        Confirmed  Deaths  Recovered  
0               0       0        0.0  
1               0       0        0.0  
2               0       0        0.0  
3               0       0        0.0  
4               0       0        0.0  
...           ...     ...        ...  
101278       1551      35     1430.0  
101279     156393    1803   146279.0  
101280       2119     615     1424.0  
101281      47622     672    40362.0  
101282      32004    1103    23413.0  

[101283 rows x 8 columns]

#### Datum als Datetime

In [30]:
full_table['Date'] = pd.to_datetime(full_table['Date'])
full_table.head()

Province/State Country/Region       Lat       Long       Date  Confirmed  \
0            NaN    Afghanistan  33.93911  67.709953 2020-01-22          0   
1            NaN        Albania  41.15330  20.168300 2020-01-22          0   
2            NaN        Algeria  28.03390   1.659600 2020-01-22          0   
3            NaN        Andorra  42.50630   1.521800 2020-01-22          0   
4            NaN         Angola -11.20270  17.873900 2020-01-22          0   

   Deaths  Recovered  
0       0        0.0  
1       0        0.0  
2       0        0.0  
3       0        0.0  
4       0        0.0

#### missing NaN 

In [31]:
full_table.isna().sum()

Province/State    70119
Country/Region        0
Lat                 371
Long                371
Date                  0
Confirmed             0
Deaths                0
Recovered          7791
dtype: int64

#### recovered mit nullen füllen

In [32]:
full_table['Recovered'] = full_table['Recovered'].fillna(0)

#### Kreuzfahrtschiffe rausnehmen

In [33]:
ship_rows = full_table['Province/State'].str.contains('Grand Princess') | full_table['Province/State'].str.contains('Diamond Princess') | full_table['Country/Region'].str.contains('Diamond Princess') | full_table['Country/Region'].str.contains('MS Zaandam')
full_ship = full_table[ship_rows]

full_table = full_table[~(ship_rows)]

### Neue Spalte `active`

In [34]:
# Active Case = confirmed - deaths - recovered
full_table['Active'] = full_table['Confirmed'] - full_table['Deaths'] - full_table['Recovered']

/opt/tljh/user/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


### Länder zusammenfassen (falls mehrere Province/State) infos vorhanden

In [35]:
full_grouped = full_table.groupby(['Date', 'Country/Region'])['Confirmed', 'Deaths', 'Recovered', 'Active'].sum().reset_index()
full_grouped

/opt/tljh/user/lib/python3.7/site-packages/ipykernel_launcher.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  """Entry point for launching an IPython kernel.


Date      Country/Region  Confirmed  Deaths  Recovered  Active
0     2020-01-22         Afghanistan          0       0        0.0     0.0
1     2020-01-22             Albania          0       0        0.0     0.0
2     2020-01-22             Algeria          0       0        0.0     0.0
3     2020-01-22             Andorra          0       0        0.0     0.0
4     2020-01-22              Angola          0       0        0.0     0.0
...          ...                 ...        ...     ...        ...     ...
70485 2021-01-26             Vietnam       1551      35     1430.0    86.0
70486 2021-01-26  West Bank and Gaza     156393    1803   146279.0  8311.0
70487 2021-01-26               Yemen       2119     615     1424.0    80.0
70488 2021-01-26              Zambia      47622     672    40362.0  6588.0
70489 2021-01-26            Zimbabwe      32004    1103    23413.0  7488.0

[70490 rows x 6 columns]

### neue Spalten New cases, New deaths and New recovered

In [36]:
import numpy as np 

# new cases 
temp = full_grouped.groupby(['Country/Region', 'Date', ])['Confirmed', 'Deaths', 'Recovered']
temp = temp.sum().diff().reset_index()
mask = temp['Country/Region'] != temp['Country/Region'].shift(1)
temp.loc[mask, 'Confirmed'] = np.nan
temp.loc[mask, 'Deaths'] = np.nan
temp.loc[mask, 'Recovered'] = np.nan
# renaming columns
temp.columns = ['Country/Region', 'Date', 'New cases', 'New deaths', 'New recovered']
# merging new values
full_grouped = pd.merge(full_grouped, temp, on=['Country/Region', 'Date'])
# filling na with 0
full_grouped = full_grouped.fillna(0)
# fixing data types
cols = ['New cases', 'New deaths', 'New recovered']
full_grouped[cols] = full_grouped[cols].astype('int')
# 
full_grouped['New cases'] = full_grouped['New cases'].apply(lambda x: 0 if x<0 else x)

/opt/tljh/user/lib/python3.7/site-packages/ipykernel_launcher.py:4: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  after removing the cwd from sys.path.


In [27]:
full_grouped.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 69920 entries, 0 to 69919
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Date            69920 non-null  datetime64[ns]
 1   Country/Region  69920 non-null  object        
 2   Confirmed       69920 non-null  int64         
 3   Deaths          69920 non-null  int64         
 4   Recovered       69920 non-null  float64       
 5   Active          69920 non-null  float64       
 6   New cases       69920 non-null  int64         
 7   New deaths      69920 non-null  int64         
 8   New recovered   69920 non-null  int64         
dtypes: datetime64[ns](1), float64(2), int64(5), object(1)
memory usage: 5.3+ MB


In [37]:
full_grouped["Recovered"].astype(int)
full_grouped["Active"].astype(int)
full_grouped.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 70490 entries, 0 to 70489
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Date            70490 non-null  datetime64[ns]
 1   Country/Region  70490 non-null  object        
 2   Confirmed       70490 non-null  int64         
 3   Deaths          70490 non-null  int64         
 4   Recovered       70490 non-null  float64       
 5   Active          70490 non-null  float64       
 6   New cases       70490 non-null  int64         
 7   New deaths      70490 non-null  int64         
 8   New recovered   70490 non-null  int64         
dtypes: datetime64[ns](1), float64(2), int64(5), object(1)
memory usage: 5.4+ MB


In [39]:
full_grouped.to_csv("WHO_Corona_global.csv", index=False)